File: zeroday-server/src/main/kotlin/com/zeroday/handler/handlers/SkillHandler.kt
package com.zeroday.handler.handlers

import com.zeroday.handler.HandlerContext
import com.zeroday.handler.HandlerResult
import com.zeroday.handler.MessageHandler
import com.zeroday.handler.reqStr
import com.zeroday.model.SkillTree
import com.zeroday.protocol.ErrorCategory
import com.zeroday.protocol.RequestTypes
import com.zeroday.protocol.ResponseTypes
import com.zeroday.protocol.ServerError
import com.zeroday.service.SkillService
import kotlinx.serialization.json.JsonObject

class SkillHandler(
    private val skillService: SkillService
) : MessageHandler {
    override val handledTypes: Set<String> = setOf(
        RequestTypes.GET_SKILL_TREE,
        RequestTypes.UNLOCK_SKILL
    )

    override suspend fun handle(ctx: HandlerContext, type: String, payload: JsonObject): HandlerResult {
        val playerId = ctx.authenticatedPlayerId()
            ?: return HandlerResult.Err(ServerError.fromMessage("Authentication required", ErrorCategory.AUTHENTICATION))
        return when (type) {
            RequestTypes.GET_SKILL_TREE -> tree(ctx, playerId)
            RequestTypes.UNLOCK_SKILL -> unlock(ctx, playerId, payload)
            else -> HandlerResult.Err(ServerError.fromMessage("Unsupported skill request: $type", ErrorCategory.VALIDATION))
        }
    }

    private suspend fun tree(ctx: HandlerContext, playerId: String): HandlerResult {
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        val m = skillService.effectiveMultipliers(player)
        val grouped = SkillTree.values().associateWith { skillService.byTree(it) }
        val trees = grouped.map { (tree, defs) ->
            mapOf(
                "tree" to tree.name,
                "skills" to defs.map { def ->
                    val unlocked = player.unlockedSkills.contains(def.id)
                    val canUnlock = !unlocked &&
                        player.skillPoints >= def.cost &&
                        def.prerequisites.all { player.unlockedSkills.contains(it) }
                    mapOf(
                        "id" to def.id,
                        "name" to def.name,
                        "description" to def.description,
                        "tier" to def.tier,
                        "cost" to def.cost,
                        "prerequisites" to def.prerequisites,
                        "unlocked" to unlocked,
                        "can_unlock" to canUnlock,
                        "effects" to def.effects.map { effect ->
                            mapOf("kind" to effect.kind, "params" to effectParams(effect))
                        }
                    )
                }
            )
        }
        return HandlerResult.Ok(
            type = ResponseTypes.SKILL_TREE,
            payload = mapOf(
                "trees" to trees,
                "skill_points" to player.skillPoints,
                "unlocked_count" to player.unlockedSkills.size,
                "multipliers" to mapOf(
                    "credits_earned" to m.creditsEarned,
                    "xp_earned" to m.xpEarned,
                    "resource_cost" to m.resourceCost,
                    "regen_rate" to m.regenRate,
                    "bonus_cpu" to m.bonusCpu,
                    "bonus_ram" to m.bonusRam,
                    "bonus_bandwidth" to m.bonusBandwidth
                )
            )
        )
    }

    private suspend fun unlock(ctx: HandlerContext, playerId: String, payload: JsonObject): HandlerResult {
        val skillId = try { payload.reqStr("skill_id") } catch (_: Exception) {
            return HandlerResult.Err(ServerError.fromMessage("Skill ID required", ErrorCategory.VALIDATION))
        }
        val player = ctx.services.playerService.getPlayer(playerId)
            ?: return HandlerResult.Err(ServerError.fromMessage("Player not found", ErrorCategory.NOT_FOUND))
        return when (val result = skillService.unlock(player, skillId)) {
            is SkillService.UnlockResult.Success -> HandlerResult.Ok(
                type = ResponseTypes.SKILL_UNLOCKED,
                payload = mapOf(
                    "skill_id" to result.skill.id,
                    "name" to result.skill.name,
                    "skill_points_remaining" to player.skillPoints
                )
            )
            is SkillService.UnlockResult.AlreadyOwned -> HandlerResult.Ok(
                type = ResponseTypes.SKILL_UNLOCKED,
                payload = mapOf(
                    "skill_id" to result.skill.id,
                    "name" to result.skill.name,
                    "already_owned" to true,
                    "skill_points_remaining" to player.skillPoints
                )
            )
            is SkillService.UnlockResult.InsufficientPoints -> HandlerResult.Err(
                ServerError.fromMessage(
                    "Need ${result.cost} skill points (have ${result.available})",
                    ErrorCategory.VALIDATION
                )
            )
            is SkillService.UnlockResult.MissingPrereqs -> HandlerResult.Err(
                ServerError.fromMessage(
                    "Missing prerequisite skills: ${result.missing.joinToString()}",
                    ErrorCategory.VALIDATION
                )
            )
            is SkillService.UnlockResult.Unknown -> HandlerResult.Err(
                ServerError.fromMessage("Unknown skill: ${result.id}", ErrorCategory.NOT_FOUND)
            )
        }
    }

    private suspend fun effectParams(effect: com.zeroday.model.SkillEffect): Map<String, Any?> = when (effect) {
        is com.zeroday.model.SkillEffect.MultiplyCreditsEarned -> mapOf("factor" to effect.factor)
        is com.zeroday.model.SkillEffect.MultiplyXpEarned -> mapOf("factor" to effect.factor)
        is com.zeroday.model.SkillEffect.AddMaxResource -> mapOf("cpu" to effect.cpu, "ram" to effect.ram, "bandwidth" to effect.bandwidth)
        is com.zeroday.model.SkillEffect.ReduceResourceCost -> mapOf("factor" to effect.factor)
        is com.zeroday.model.SkillEffect.IncreaseRegenRate -> mapOf("factor" to effect.factor)
    }
}

File: zero-server-csharp/src/ZeroDayMMO.Server/Handlers/SkillHandler.cs
using System.Text.Json;
using ZeroDayMMO.Server.Services;
using ZeroDayMMO.Shared;
using ZeroDayMMO.Shared.Models;

namespace ZeroDayMMO.Server.Handlers;

public class SkillHandler : IHandler
{
    private readonly IPlayerService _playerService;
    private readonly SkillService _skillService;

    public string MessageType => "get_skill_tree";

    public SkillHandler(IPlayerService playerService, SkillService skillService)
    {
        _playerService = playerService;
        _skillService = skillService;
    }

    public async Task<IActionResult> HandleAsync(string connectionId, JsonElement? payload, string? requestId)
    {
        if (payload is null)
            return new MessageResult("error", connectionId, new { message = "Missing payload" });

        var type = payload.Value.Str("type") ?? "get_skill_tree";

        return type switch
        {
            "get_skill_tree" => await Tree(connectionId),
            "unlock_skill" => await Unlock(connectionId, payload.Value),
            _ => new MessageResult("error", connectionId, new { message = $"Unsupported skill request: {type}" })
        };
    }

    private async Task<IActionResult> Tree(string connectionId)
    {
        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        var m = _skillService.GetEffectiveMultipliers(player);
        var grouped = Enum.GetValues<SkillTree>().ToDictionary(tree => tree, tree => _skillService.ByTree(tree));

        var trees = grouped.Select(kvp =>
        {
            var tree = kvp.Key;
            var defs = kvp.Value;
            return new
            {
                tree = tree.ToString(),
                skills = defs.Select(def =>
                {
                    var unlocked = player.UnlockedSkills.Contains(def.Id);
                    var canUnlock = !unlocked &&
                        player.SkillPoints >= def.Cost &&
                        def.Prerequisites.All(p => player.UnlockedSkills.Contains(p));

                    return new
                    {
                        id = def.Id,
                        name = def.Name,
                        description = def.Description,
                        tier = def.Tier,
                        cost = def.Cost,
                        prerequisites = def.Prerequisites,
                        unlocked,
                        can_unlock = canUnlock,
                        effects = def.Effects.Select(EffectParams).ToList()
                    };
                }).ToList()
            };
        }).ToList();

        return new MessageResult("skill_tree", connectionId, new
        {
            trees,
            skill_points = player.SkillPoints,
            unlocked_count = player.UnlockedSkills.Count,
            multipliers = new
            {
                credits_earned = m.CreditsEarned,
                xp_earned = m.XpEarned,
                resource_cost = m.ResourceCost,
                regen_rate = m.RegenRate,
                bonus_cpu = m.BonusCpu,
                bonus_ram = m.BonusRam,
                bonus_bandwidth = m.BonusBandwidth
            }
        });
    }

    private async Task<IActionResult> Unlock(string connectionId, JsonElement payload)
    {
        string skillId;
        try
        {
            skillId = payload.ReqStr("skill_id");
        }
        catch
        {
            return new MessageResult("error", connectionId, new { message = "Skill ID required" });
        }

        var player = _playerService.GetPlayer(connectionId);
        if (player is null)
            return new MessageResult("error", connectionId, new { message = "Player not found" });

        var result = _skillService.Unlock(player, skillId);

        return result switch
        {
            UnlockResult.Success s => new MessageResult("skill_unlocked", connectionId, new
            {
                skill_id = s.Skill.Id,
                name = s.Skill.Name,
                skill_points_remaining = player.SkillPoints
            }),
            UnlockResult.AlreadyOwned a => new MessageResult("skill_unlocked", connectionId, new
            {
                skill_id = a.Skill.Id,
                name = a.Skill.Name,
                already_owned = true,
                skill_points_remaining = player.SkillPoints
            }),
            UnlockResult.InsufficientPoints i => new MessageResult("error", connectionId, new
            {
                message = $"Need {i.Cost} skill points (have {i.Available})"
            }),
            UnlockResult.MissingPrereqs m => new MessageResult("error", connectionId, new
            {
                message = $"Missing prerequisite skills: {string.Join(", ", m.Missing)}"
            }),
            UnlockResult.Unknown u => new MessageResult("error", connectionId, new
            {
                message = $"Unknown skill: {u.Id}"
            }),
            _ => new MessageResult("error", connectionId, new { message = "Unknown error" })
        };
    }

    private static object EffectParams(SkillEffect effect) => effect switch
    {
        MultiplyCreditsEarned mc => new { kind = "multiply_credits_earned", factor = mc.Factor },
        MultiplyXpEarned mx => new { kind = "multiply_xp_earned", factor = mx.Factor },
        AddMaxResource ar => new { kind = "add_max_resource", cpu = ar.Cpu, ram = ar.Ram, bandwidth = ar.Bandwidth },
        ReduceResourceCost rc => new { kind = "reduce_resource_cost", factor = rc.Factor },
        IncreaseRegenRate ir => new { kind = "increase_regen_rate", factor = ir.Factor },
        _ => new { kind = "unknown" }
    };
}
